File: Difficulty: Easy/Chocolate Distribution Problem/chocolate-distribution-problem.js
//{ Driver Code Starts
// Initial Template for javascript

'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => { inputString += inputStdin; });

process.stdin.on('end', _ => {
    inputString =
        inputString.trim().split('\n').map(string => { return string.trim(); });

    main();
});

function readLine() { return inputString[currentLine++]; }

function printList(res, n) {
    let s = "";
    for (let i = 0; i < n; i++) {
        s += res[i];
        s += " ";
    }
    console.log(s);
}

function main() {
    let t = parseInt(readLine());
    let i = 0;
    for (; i < t; i++) {

        let input_ar1 = readLine().split(' ').map(x => parseInt(x));
        let n = input_ar1.length;
        let arr = new Array(n);
        let m = parseInt(readLine());
        for (let i = 0; i < n; i++) {
            arr[i] = input_ar1[i];
        }
        let obj = new Solution();
        let res = obj.findMinDiff(arr, m);
        console.log(res);
        console.log("~");
    }
} // } Driver Code Ends

// } Driver Code Ends

// User function Template for JavaScript

/**
 * @param {number[]} arr - The input array of integers
 * @param {number} m - The number of students (size of subset)
 * @returns {number} - The minimum difference between maximum and minimum chocolates
 */

class Solution {
    findMinDiff(arr, m) {
        // Sort the array to arrange chocolate packets in ascending order
        arr.sort((a, b) => a - b);

        let minDiff = Infinity; // Initialize min difference to a large value

        // Iterate through possible windows of size 'm'
        for (let i = 0; i + m - 1 < arr.length; i++) {
            let diff = arr[i + m - 1] - arr[i]; // Compute difference between max & min in the window
            minDiff = Math.min(minDiff, diff); // Update minDiff if the current difference is smaller
        }

        return minDiff; // Return the minimum difference found
    }
}

File: Difficulty: Easy/Chocolate Distribution Problem/chocolate-distribution-problem.cpp
//{ Driver Code Starts
// Initial Template for C++

#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends

//Back-end complete function Template for C++

int minDiff(vector<int> &arr, int k) {
    // Sort the array to calculate the minimum difference between the k largest and smallest elements
    sort(arr.begin(), arr.end());

    // Calculate the initial difference between the first k elements
    int minDiff = arr[k-1] - arr[0];

    // Iterate over the array to find the minimum difference for any subset of k elements
    for(int i = k; i < arr.size(); i++) {
        // Update the minimum difference
        minDiff = min(minDiff, arr[i] - arr[i-k+1]);
    }

    // Return the minimum difference found
    return minDiff;
}



//{ Driver Code Starts.

vector<int> inputLine() {
    string str;
    getline(cin, str);
    stringstream ss(str);
    int num;
    vector<int> res;
    while (ss >> num) {
        res.push_back(num);
    }
    return res;
}

int main() {

    int t;
    cin >> t;
    cin.ignore();
    while (t--) {
        vector<int> arr = inputLine();

        int m;
        cin >> m;
        cin.ignore();

        cout << minDiff(arr, m) << endl;
        cout << "~" << endl;
    }
    return 0;
}
// } Driver Code Ends